File: src/engine/event_bus.py
class EventBus:
    def __init__(self):
        self._listeners = {}

    def subscribe(self, event_type, callback, priority=0):
        """
        Subscribe a callback to an event.
        Higher priority callbacks will run first.
        """
        if event_type not in self._listeners:
            self._listeners[event_type] = []
        self._listeners[event_type].append((priority, callback))
        # Sort in descending order of priority
        self._listeners[event_type].sort(key=lambda x: x[0], reverse=True)

    def unsubscribe(self, event_type, callback):
        """Unsubscribe a callback from an event."""
        if event_type in self._listeners:
            self._listeners[event_type] = [
                item for item in self._listeners[event_type] if item[1] != callback
            ]

    def publish(self, event_type, *args, **kwargs):
        """Publish an event, calling all registered callbacks in order of priority."""
        if event_type in self._listeners:
            # Create a copy of the list in case listeners mutate subscriptions during callback execution
            for _, callback in list(self._listeners[event_type]):
                callback(*args, **kwargs)

    def clear(self):
        """Clear all registered event listeners."""
        self._listeners.clear()


# Global Event Bus instance
bus = EventBus()

File: love2d/src/engine/event_bus.lua
local EventBus = {}
EventBus.__index = EventBus

function EventBus.new()
    local self = setmetatable({}, EventBus)
    self._listeners = {}
    return self
end

function EventBus:subscribe(event_type, callback, priority)
    priority = priority or 0
    if not self._listeners[event_type] then
        self._listeners[event_type] = {}
    end
    table.insert(self._listeners[event_type], {callback = callback, priority = priority})
    -- Sort in descending order of priority
    table.sort(self._listeners[event_type], function(a, b)
        return a.priority > b.priority
    end)
end

function EventBus:unsubscribe(event_type, callback)
    if not self._listeners[event_type] then return end
    local newList = {}
    for _, item in ipairs(self._listeners[event_type]) do
        if item.callback ~= callback then
            table.insert(newList, item)
        end
    end
    self._listeners[event_type] = newList
end

function EventBus:publish(event_type, ...)
    if not self._listeners[event_type] then return end
    local listCopy = {}
    for _, item in ipairs(self._listeners[event_type]) do
        table.insert(listCopy, item.callback)
    end
    
    for _, callback in ipairs(listCopy) do
        callback(...)
    end
end

function EventBus:clear()
    self._listeners = {}
end

local bus = EventBus.new()

return {
    EventBus = EventBus,
    bus = bus
}
